### ISSUE TITLE
Bug: ProjectionPlot line annotation modifies array

### ISSUE THREAD

**Issue by claytonstrawn** [yt-project/yt#3449]
### Bug report

When I use annotate_line, it modifies the original array of startpoints and endpoints, and doesn't keep a consistent size. It has something to do with "unitary" from the UniformGridData unit_registry

**Code for reproduction**

```python
import yt
yt.funcs.mylog.setLevel(50)

data_array = np.random.random((10,10,10))
bbox = np.array([[-100,100],[-100,100],[-100,100]])
data = {('gas','density'):(data_array, 'g*cm**(-3)')}
ds = yt.load_uniform_grid(data, data_array.shape, length_unit="kpc", bbox=bbox)

start_source = np.array((0,0,-0.5))
end_source =  np.array((0,0,0.5))
start = ds.arr(start_source,'unitary')
end = ds.arr(end_source,'unitary')

p = yt.ProjectionPlot(ds,0,'number_density')
p.annotate_line(start,end)
print('\n before')
print("start",start,type(start))
print("end",end,type(end))
print("start_source",start_source,type(start_source))
print("end_source",end_source,type(end_source))
p.show()
print('\n after')
print("start",start,type(start))
print("end",end,type(end))
print("start_source",start_source,type(start_source))
print("end_source",end_source,type(end_source))

start = ds.arr(start,'unitary')
end = ds.arr(end,'unitary')
print('\n trying to coerce')
print("start",start,type(start))
print("end",end,type(end))
print("start_source",start_source,type(start_source))
print("end_source",end_source,type(end_source))
```

**Actual outcome**

```
 before
start [ 0.   0.  -0.5] unitary <class 'unyt.array.unyt_array'>
end [0.  0.  0.5] unitary <class 'unyt.array.unyt_array'>
start_source [ 0.   0.  -0.5] <class 'numpy.ndarray'>
end_source [0.  0.  0.5] <class 'numpy.ndarray'>
Image

 after
start [   0.    0. -100.] code_length <class 'unyt.array.unyt_array'>
end [  0.   0. 100.] code_length <class 'unyt.array.unyt_array'>
start_source [   0.    0. -100.] <class 'numpy.ndarray'>
end_source [  0.   0. 100.] <class 'numpy.ndarray'>

 trying to coerce
start [   0.    0. -100.] unitary <class 'unyt.array.unyt_array'>
end [  0.   0. 100.] unitary <class 'unyt.array.unyt_array'>
start_source [   0.    0. -100.] <class 'numpy.ndarray'>
end_source [  0.   0. 100.] <class 'numpy.ndarray'>
```

**Expected outcome**

We would expect the start and end arrays to simply remain in `unitary` units, or if they changed to `code_length`, at least be able to recover the unitary array by explicitly reassigning it with `ds.arr` again. Most importantly, it should _not_ modify the original array `start_source`, because that is just a numpy array somewhere, not a unyt array, so it definitely should not be modified by "unit conversion"

**Version Information**

yt is installed from source, with version `'4.0.dev0'`

**Comment by matthewturk** [comment]
I agree, this is not desired behavior.

I'm digging into this, and I *think* that it's possible the change is happening inside the function `_convert_to_plot`, where the unyt method `to` is called.  Can you try changing that, which is on line 143, to `in_units`?

**Comment by claytonstrawn** [comment]
It doesn't look like that change affected the outcomes here. It looks to me like there's two things I think weren't supposed to happen.

1. That unyt is allowed to change existing non-unyt numpy arrays at all (i.e. I think start_source should never be changed from 0.5 to 100). Should it explicitly make a copy at some point instead of just assigning a numpy array as a value?
2. When I create it again with ds.arr(start,'unitary'), it should recognize that it is already a unyt array and just convert itself. In testing that now though I'm realizing maybe this is not standard behavior, I just assumed unyt would do that. I would have thought a2 == a1.in_units('kpc') below.
```
a1 = ds.arr([0.5,0.5,0.5],'unitary')
a2 = ds.arr(a1,'kpc')
print("a1:%s, a2:%s"%(a1,a2))
>a1:[0.5 0.5 0.5] unitary, a2:[0.5 0.5 0.5] kpc
```
I don't think the actual conversion is wrong, just doing it multiple times was messing up the positions of the sightlines.

**Comment by claytonstrawn** [comment]
@chrishavlin Ok, that was my mistake on the new array behavior. 

Your solution worked, though! It no longer modifies the original array in the notebook. I could do a pull request with that change if needed.

On another note, I think from the user end, I've never really wanted `convert_to_units` to modify my arrays in place, and I've had a couple bugs with it when I've had that happen inside the yt backend. I wonder if it would be possible to have it do in-place conversions only when told explicitly (because I'm sure that's much better for efficiency in certain contexts), and otherwise create new arrays? Idk, just a thought.

**Comment by matthewturk** [comment]
@claytonstrawn I think the main idea for conversion in place is to reduce memory overhead, *but*, my experience completely aligns with yours.

Something we've talked about in the past has been to make a class of "immutable units" for unyt arrays (I think we called them `IndexArrays` at one point) which would always be in the same units, so you wouldn't need to (for instance) check that they're in `code_units` inside a tight loop or something.

**Comment by chrishavlin** [comment]
> Your solution worked, though! It no longer modifies the original array in the notebook. I could do a pull request with that change if needed.

@claytonstrawn, sure! go for it! I agree that the present behavior is not ideal.

Some discussion may come up in the pull request in terms of the memory efficiency points you mention. But given that `_sanitize_coord_system` does the following

```python
            coord = self._project_coords(plot, coord)  # uses in place modification with `convert_to_units`
            coord = self._convert_to_plot(plot, coord)  uses `to` to create a new copy 
```
it's clearly not being efficient at present... there might be an argument for also switching `_convert_to_plot` to use `convert_to_units` after changing `_project_coords` to use `to`  to avoid copying the array twice, but I think starting with a PR with just the `_project_coords` change is a good start.

**Comment by chummels** [comment]
Just to chime in here, I *do* think there is a reason for `convert_to_units()` to behave the way it currently does.  While most cases can appropriate be covered with `array.to('units')` to return a duplicate copy of `array` in the desired units, there are definitely cases that I've encountered where you need to intrinsically change `array` in place, so future operations on that array are in the desired units.  I'll try to think of them to give you an example.

**Comment by neutrinoceros** [comment]
Fixed via #3463
